### ISSUE TITLE
Jumpy/stuttering rendering when dynamic shadows are enabled (high dtime jitter, but good average FPS)

### ISSUE THREAD

**Issue by VanessaE** [minetest/minetest#11357]
##### Minetest version
Minetest engine/client is at git commit dc165fe9, master branch, and was built thusly:

<details><summary>[Show build script...]</summary>

```bash
#!/bin/bash

# If needed, compile and install MT Irrlicht
# ~/Scripts/build-Irrlicht-MT.sh

pushd .

export CXXFLAGS="-O3"

cd /home/vanessa/Minetest-related/minetest-engine
rm -f CMakeCache.txt

make clean
cmake . -DRUN_IN_PLACE=0 -DCMAKE_BUILD_TYPE=Release \
	-DENABLE_FREETYPE=true -DENABLE_LEVELDB=true \
	-DENABLE_GETTEXT=true \
	-DCMAKE_INSTALL_PREFIX=/usr \
	-DIRRLICHT_LIBRARY=/usr/lib/x86_64-linux-gnu/libIrrlichtMt.so \
	-DIRRLICHT_INCLUDE_DIR=/usr/include/irrlichtmt/
make -j8

# install them

echo "sudo make install..."
sudo make install

popd
```
</details>

MInetest-Irrlicht is at git commit `d7127df9`, master branch, and was built thusly:
<details><summary>[Show build script...]</summary>

```bash
#!/bin/bash

pushd .

export CXXFLAGS="-O3"

# Build MT's fork of irrlicht first

cd /home/vanessa/Minetest-related/irrlicht-mt-fork/
rm -f CMakeCache.txt

cmake . -DBUILD_SHARED_LIBS=ON -DCMAKE_INSTALL_PREFIX=/usr
make clean
make -j8
echo "sudo make install irrlicht..."
sudo make install

popd
```
</details>

##### OS / Hardware

Operating system:  Debian "bullseye"
CPU: Phenom II X6 1055T, 2.8 GHz

GPU model: AMD R9 280X (official `amdgpu` driver)
OpenGL version: 4.6
Desktop environment:  XFCE 4.16

##### Summary

When dynamic shadows are enabled, rendering gets very jumpy/stuttery as more and more terrain is loaded, such as when running at top speed down a long road, and it doesn't seem to matter how much is actually visible in front of me due to my view range (as set by [keypad +/-]).  Even if I turn my view range down to, say, 20m, the rendering stutter remains.

Important:  I am *not* talking about the lag that happens as new mapblocks are loaded from the world or server, that's an entirely different issue.  

I might get an average of 30 FPS in a given scene, but the framerate will bounce wildly between say, 20 and 40, causing the view to stutter as I walk around.  Indeed, at its worst, I've seen dtime jitter of over 300%.  If it gets too much over about 150%, the game is simply unplayable.

If I leave the game sitting for a while (a few minutes), the stutter calms down for a short time.  The effect suggests that a cache or buffer is running dry while I'm wandering around, and gradually fills back up while I'm standing still (i.e. when full, performance improves).  This may or may not require waiting in the 'pause' menu, I didn't test if simply stepping away without pausing was enough.  Since neither my CPU nor GPU show indications of overheat or overload, according to the widgets in my desktop panel/taskbar, I'm guessing hardware is not the problem.

It does not matter how long the terrain has been loaded -- if the stutter has calmed down and I start running around within terrain I know for sure is already loaded, for example going back down the aforementioned long road, then going back the other way again, it'll start stuttering again after a few seconds.

Dreambuilder may be required to create the effect, as a matter of how much stuff it adds.

It seems to depend slightly on the number of visible artificial lights (for example, `ilights` mod), but I'm not entirely sure about this.

It's worse when using HDX 256.

It's more obvious when "cinematic" mode is enabled, though it may not necessarily actually be worse.

If I turn off the dynamic shadows, things return to normal.

##### Steps to reproduce

I can't give a predictable, solid list of steps, because I'm not clear on what would be a minimal test case, but maybe try this:

* Set your shader settings thus:
  ```
  enable_waving_water = true
  enable_waving_plants = true
  enable_waving_leaves = false
  
  enable_dynamic_shadows = true
  shadow_map_color = true
  shadow_filters = 1
  shadow_levels = 0
  shadow_map_texture_size = 2048
  shadow_strength = 0.4
  shadow_map_texture_32bit = true
  shadow_map_max_distance = 350
  shadow_soft_radius = 2
  ```
* Enable cinematic mode (`cinematic = true` in minetest.conf).
* Enable HDX, 256 px resolution -- make sure you have the full pack, not the minimal mtg-only version.
* Join my "Traintable" server: `daconcepts.com:30004`.
  Warning: this server runs Dreambuilder and between that and full HDX 256, you will need no less than 8 GB of *free* RAM.
* Increase your view range to around 100-150 meters -- just whatever is enough to stress your system a bit and bring the framerate down into the 20-30 FPS range.
* Set fast mode (`J`), and run east from the spawn.
* When you come to the tunnel, veer left to stay on the surface road.
* Stop when you get to where the streetlights stop.  At the time of this report, the road ends in a heavily-forested area with no player structures.
* Stand in place and pan the camera around slowly, try to avoid looking where little terrain has been loaded yet.

By the time you get even half-way down that road, you should notice the FPS stutter, but standing there looking around should surely make it show up.

**Comment by VanessaE** [comment]
I recorded a short video showing the stutter:

https://youtu.be/IJKdsh6RxMc

Not sure why, but it's a bit less severe here compared to earlier.

**Comment by lhofhansl** [comment]
Hmm... I'd expect shadow to create more load on both CPU and GPU. I found especially colored shadows do that.

What if you change the shadow map update time from the 0.2s default to something lower or higher, do you see a change in the stutter?

**Comment by 0xLiso** [comment]
I'm not sure, but i feel the problem is the amount of map SM renders. I'm working on it.

**Comment by VanessaE** [comment]
I tested @0xLiso's `dev` branch earlier today (about 6½ hours prior to this comment) and this issue is massively improved there.

**Comment by hecktest** [comment]
Shadows are rendered only every fraction of a second as an "optimization" which does nothing to improve the worst case frame time, only makes it affect every n-th frame.

It's basically cheating the FPS counter. Also causes visual artifacts when digging and building. And how does it work with CAOs if they're clearly updated every frame? @0xLiso , don't tell me you allocate a second shadow map just for this...

**Comment by 0xLiso** [comment]
> Shadows are rendered only every fraction of a second as an "optimization" which does nothing to improve the worst case frame time, only makes it affect every n-th frame.

> It's basically cheating the FPS counter. Also causes visual artifacts when digging and building. And how does it work with CAOs if they're clearly updated every frame? @0xLiso , don't tell me you allocate a second shadow map just for this...


No, Map shadows are rendered every fraction of a second **because the map is updated every fraction of a second**. And due the slowness of sending thousands of extra drawcalls, I decided to create an extra RTT texture to avoid those thousands of drawcalls every frame. 
 

Yes @kilbith pointed those visual artifacts time ago, so we created a configuration  parameter to tune it. It is called "shadow_update_time" if you want to test it.

**Comment by hecktest** [comment]
But is the update actually distributed over several frames or not?

**Comment by 0xLiso** [comment]
> Map shadows are rendered every fraction of a second **because the map is updated every fraction of a second**.

So, yes. Shadow's drawlist is update only when map updates and also it's rendered that moment.

**Comment by hecktest** [comment]
I'm not asking if it's done in intervals. I'm asking whether the update is actually budgeted over several frames, because OP's problem suggests that it is not.

**Comment by 0xLiso** [comment]
Sorry, I'm not getting what you are asking.

The shadow pass does this:

1. Render ClientMap's depth every time that ClientMap updates to avoid excessive drawcalls. 
1.1 To minimize the visual artifacts when digging and building , this interval was set as configuration parameter. (client/game.cpp line 3840 )
1.2  This depth render is done in a RTT only when ClientMap is updated.
2. Render every frame depth for all dynamic entities (CAO) in other texture
3. Mix both depth textures in the Final Shadow Texture.
3.1 This mix pass is also used to "pack" both shadow and "color shadow" in the same texture, using:
    - r -> for solid elements depth 
    - g -> 0 but it can be used in a future for other techniques (maybe VMS?)  (only if color shadows if active)
    - b -> for translucent elements depth (only if color shadows if active)
    - a -> for translucent packed color (only if color shadows if active)

This pipeline was implemented initially thinking in CSM for which we needed 3 textures for shadows splits. But at that time MT only was able to handle 4 textures ( Now with IrrlitchMT supports up to 8 ). 
And from those 4 textures 3 of them were already used ( 1º base texture, 2ª properties texture and 3th for normal texture ).
So we only had 1 texture left. And that is the reason I had to make originally a second pass to mix the shadow textures. 

When I switched to PSM I reused that mix pass to mix clientmap shadow texture and CAO shadow texture.

I'm sure all of this is wrong and It could be done in a incredible better way, but it is what I did with the knowledge and the constrains I had.

**Comment by hecktest** [comment]
I don't really see a noticeable performance difference between an interval of 0.001 and 0.2, only artifacts. Might as well just make this a boolean setting - either update it all immediately after the ClientMap update, or peg it to a full ClientMap update interval, budgeting the shadow update over this time. That way it'll actually smooth the framerate out.

_but_

Delays of any kind conflict with cascaded shadow mapping (because it is frustum based), so once the renderer is unclogged enough to allow CSM again, this hack might have to go away entirely.

**Comment by x2048** [comment]
@VanessaE @hecktest Now with #11422 merged, and shadow map update is spread over multiple frames (that can be controlled), can we close with bug?
I've checked at TrainTable server (featured in the video), and I can still notice stutter which I think is caused by draw list and shadow draw list update and the occlusion culling algorithm there, not by rendering the SM texture(s).

**Comment by VanessaE** [comment]
To be sure I got up-to-date, clean code, I did `git reset --hard HEAD~200; git pull`, then built and tested.

There's some very minor stuttering still, but I don't think it's related to the shadows.  Rather, I think we're just down to the usual overall mediocre rendering performance (which I guess is what you meant by the "draw list" stuff).

Overall, its WAY better, even with HDX enabled, enough-so that I'd consider this issue as fixed.
